### PR TITLE
[MU4] Added draw objects trace

### PR DESCRIPTION
**Pull request by igorkorsukov**
See also https://github.com/musescore/MuseScore/pull/7614

### CODE REVIEW

File: src/libmscore/draw/painter.h [hunk 1]
```diff
@@ -229,6 +233,37 @@ inline void Painter::drawText(int x, int y, const QString& text)
...
+
+#ifdef TRACE_OBJ_DRAW_ENABLED
+    #define TRACE_OBJ_DRAW \
+    mu::draw::PainterObjMarker __drawObjMarker(painter, name(), pagePos())
+
+    #define TRACE_OBJ_DRAW_C(painter, objName, objPagePos) \
+    mu::draw::PainterObjMarker __drawObjMarker(painter, objName, objPagePos)
+#else
+    #define TRACE_OBJ_DRAW
+    #define TRACE_OBJ_DRAW_C(painter, objName)
```
Comment (cbjeukendrup) [inline src/libmscore/draw/painter.h:265]: Isn't `, objPagePos` missing here?

File: src/libmscore/draw/painter.h [hunk 2]
```diff
@@ -229,6 +233,37 @@ inline void Painter::drawText(int x, int y, const QString& text)
...
     drawText(QPointF(x, y), text);
 }
+
+class PainterObjMarker
+{
+public:
+    PainterObjMarker(Painter* p, const std::string& name, const QPointF& objPagePos)
+        : m_painter(p), m_name(name), m_objPagePos(objPagePos)
+    {
+        p->beginObject(name, m_objPagePos);
```
[the remark this answers is not in the document]
Comment (igorkorsukov) [reply src/libmscore/draw/painter.h]: There is no difference :) I did it consistently.

File: src/libmscore/draw/painter.h (hunk 1 above)
Comment (igorkorsukov) [reply src/libmscore/draw/painter.h]: fixed

File: src/libmscore/draw/painter.cpp [hunk 3]
```diff
@@ -75,17 +74,34 @@ QPainter* Painter::qpainter() const
 
 bool Painter::end()
 {
-    return m_provider->end(m_name);
+    bool ok = m_provider->end(m_name);
     if (extended) {
         extended->end(m_name);
```
Comment (RomanPudashkin) [inline src/libmscore/draw/painter.cpp:79]: Should we check the result here too?

Comment (igorkorsukov) [reply src/libmscore/draw/painter.cpp:79]: > Should we check the result here too?

There is no need. "Extended" is intended primarily for automated testing. In the near future, I will show what I have planned to do :)

**Pull request merged**